File: Gold이상문제_정리/줄세우기_2252.java
package Gold이상문제_정리;

import java.io.*;
import java.util.*;

/*
3 2
1 3
2 3
 */


public class 줄세우기_2252 {
	static int N, M;
	static ArrayList<Integer>[] adj;
	static int[] degree;
	static Queue<Integer> degree_queue = new LinkedList<>();
	static StringBuilder str = new StringBuilder();

	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		degree = new int[N+1];
		adj = new ArrayList[N+1];
		for (int i = 0; i < N+1; i++) {
			adj[i] = new ArrayList<>();
		}
		for (int m = 0; m < M; m++) {
			st = new StringTokenizer(br.readLine());
			int front = Integer.parseInt(st.nextToken());
			int back = Integer.parseInt(st.nextToken());
			adj[front].add(back);
			degree[back]++;
		}

		for (int i = 1; i < N+1; i++) {
			if(degree[i] == 0) {
				str.append(i).append(" ");
				degree_queue.add(i);
			}
		}

		while (!degree_queue.isEmpty())
		{
			int now = degree_queue.poll();

			for (int next : adj[now]) {
				degree[next]--;
				if (degree[next] == 0) {
					degree_queue.add(next);
					str.append(next).append(" ");
				}
			}
		}

		System.out.println(str);
	}
}

File: 코틀린_코테스터디/week6/줄세우기_2252.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

private var N = 0
private var M: Int = 0
private lateinit var adj: Array<MutableList<Int>>
private lateinit var degree: IntArray
private var degree_queue: Queue<Int> = LinkedList()
private var str = StringBuilder()

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()
    degree = IntArray(N + 1)
    adj = Array(N + 1) { mutableListOf() }
    for (i in 0 until N + 1) {
        adj[i] = mutableListOf()
    }
    for (m in 0 until M) {
        st = StringTokenizer(br.readLine())
        val front = st.nextToken().toInt()
        val back = st.nextToken().toInt()
        adj[front].add(back)
        degree[back]++
    }
    for (i in 1 until N + 1) {
        if (degree[i] == 0) {
            str.append(i).append(" ")
            degree_queue.add(i)
        }
    }
    while (!degree_queue.isEmpty()) {
        val now = degree_queue.poll()
        for (next in adj[now]) {
            degree[next]--
            if (degree[next] == 0) {
                degree_queue.add(next)
                str.append(next).append(" ")
            }
        }
    }
    println(str)
}